packages/pigeon/lib/swift_generator.dart

```diff
@@ -673,10 +681,10 @@ private func nilOrValue<T>(_ value: Any?) -> T? {
   void _writeCreateConnectionError(Indent indent) {
     indent.newln();
     indent.writeScoped(
-        'private func createConnectionError(withChannelName channelName: String) -> FlutterError {',
+        'private func createConnectionError(withChannelName channelName: String) -> PigeonError {',
```

Comment:
what happens to existing code that throws `FlutterError`? Do we support both? or do you completely swap it to `PigeonError`.

Reply:
If someone have a code that throws `FlutterError`, they are encouraged to change it to `PigeonError`. The whole point of this PR is that packages shouldn't use `FlutterError` to conform to the `Swift.Error` protocol. I have updated the README(`packages/pigeon/README.md`) to reflect this change.

Reply:
Can you provide the names of open-source downstream projects that I can check to see if this PR breaks them?

Reply:
> Ah that will be a breaking change and may cause chaos.

Compile time breaking changes in Pigeon are fine; we do them pretty frequently. Clients of Pigeon are fully in control of when they update their Pigeon generation, and Pigeon isn't a transitive dependency in package trees so there aren't any version conflict concerns.

Reply:
oh i was referring to runtime breaking changes. let's say we have this old code: 

```
// custom code
func fooAPI() throws {
  throw FlutterError(...)
}

// generated code
func callTheFooAPI() {
  do {
    try fooAPI()
  } catch {
    if let e = error as? FlutterError {
      // encode error and send to dart side
    }
  }
}
```
Then when the customer bump the version, which doesn't handle the `FlutterError`: 
```
// custom code
func fooAPI() throws {
  throw FlutterError(...)
}

// generated code
func callTheFooAPI() {
  do {
    try fooAPI()
  } catch {
    if let e = error as? PigeonError { // <- here
      // encode error and send to dart side
    }
  }
}
```
Since no compile error, developers may not update their code.

Reply:
Okay, I'll explain this using the source code of test codes in the pigeon package.

```dart
// https://github.com/flutter/packages/blob/dd01140f470e02b06d05ffa0213c7b26e89ae4c4/packages/pigeon/pigeons/core_tests.dart#L152
@HostApi()
abstract class HostIntegrationCoreApi {
...

  /// Returns an error, to test error handling.
  Object? throwError();
}
```

will be generated like this:
```swift
// https://github.com/flutter/packages/blob/dd01140f470e02b06d05ffa0213c7b26e89ae4c4/packages/pigeon/platform_tests/test_plugin/ios/Classes/CoreTests.gen.swift#L466
protocol HostIntegrationCoreApi {
...
  /// Returns an error, to test error handling.
  func throwError() throws -> Any?
}
...
// https://github.com/flutter/packages/blob/dd01140f470e02b06d05ffa0213c7b26e89ae4c4/packages/pigeon/platform_tests/test_plugin/ios/Classes/CoreTests.gen.swift#L701
    let throwErrorChannel = FlutterBasicMessageChannel(
      name:
        "dev.flutter.pigeon.pigeon_integration_tests.HostIntegrationCoreApi.throwError\(channelSuffix)",
      binaryMessenger: binaryMessenger, codec: codec)
    if let api = api {
      throwErrorChannel.setMessageHandler { _, reply in
        do {
          let result = try api.throwError()
          reply(wrapResult(result))
        } catch {
          reply(wrapError(error))
        }
      }
    } else {
      throwErrorChannel.setMessageHandler(nil)
    }
...
```
and current implementation of `throwError` function is as follows:

```swift
// https://github.com/flutter/packages/blob/dd01140f470e02b06d05ffa0213c7b26e89ae4c4/packages/pigeon/platform_tests/test_plugin/ios/Classes/TestPlugin.swift#L52
func throwError() throws -> Any? {
  throw FlutterError(code: "code", message: "message", details: "details")
}
```

Also, current implementation of `wrapError` function is as follows:
```swift
private func wrapError(_ error: Any) -> [Any?] {
  if let flutterError = error as? FlutterError {
    return [
      flutterError.code,
      flutterError.message,
      flutterError.details,
    ]
  }
  return [
    "\(error)",
    "\(type(of: error))",
    "Stacktrace: \(Thread.callStackSymbols)",
  ]
}
```

I believe the goal of `wrapError` is converting the error as a list with 3 elements, so that dart side decode it as a `FlutterError` object. We can throw what error we want in the dart side, and we just need to extract extra information such as `code`, `message`, and `details` from the error object.

Even we convert errors in the generated swift code to use `PigeonError`, the `wrapError` function will work as expected. I forgot to add a case for the conversion of `PigeonError` in the `wrapError` function, but it is easy to fix.

This PR's `wrapError` function will be like this:
```swift
private func wrapError(_ error: Any) -> [Any?] {
  if let flutterError = error as? FlutterError {
    return [
      flutterError.code,
      flutterError.message,
      flutterError.details,
    ]
  }
  if let pigeonError = error as? PigeonError {
    return [
      pigeonError.code,
      pigeonError.message,
      pigeonError.details,
    ]
  }
  return [
    "\(error)",
    "\(type(of: error))",
    "Stacktrace: \(Thread.callStackSymbols)",
  ]
}
```

Then, changes in the generated code won't make any problem in the dart side, even some older Swift code is throwing `FlutterError` object. Of course, as I wrote in the previous comment, implementations are encouraged to use `PigeonError` instead of `FlutterError` in the future to deal with the redundant protocol conformance issue.

In short, I believe that no compatibility issue will occur even after this PR is merged.

Reply:
> oh i was referring to runtime breaking changes. let's say we have this old code:
> 
> ```
> // custom code
> func fooAPI() throws {
> [...]

Ah, right. I was thinking of the aync functions, where the type is part of the signature.

We definitely still need `FlutterError` handling in the `wrapError` function for runtime compatibility with synchronous code. That's a tiny amount of code though, luckily.